crates/ruff/src/rules/flake8_todo/rules.rs

```diff
@@ -0,0 +1,326 @@
+// ```python
+// # TODO(evanrittenhouse): This is a completely valid TODO
+// ```
+// will yield [Some("TODO") Some("evanrittenhouse"), Some(":"), Some(" "), Some("This is a completely valid TODO")], whereas
+// ```python
+// # ToDo this is completely wrong
+// ```
+// will yield [Some("ToDo"), None, None, Some(" "), Some("this is completely wrong")]. Note the
+// `Nones` for the colon and space checks.
+//
+// Note: Regexes taken from https://github.com/orsinium-labs/flake8-todos/blob/master/flake8_todos/_rules.py#L12.
+static TODO_REGEX: Lazy<Regex> = Lazy::new(|| {
+    Regex::new(r"^#\s*(?P<tag>[tT][oO][dD][oO]|BUG|FIXME|XXX)(\(.*\))?(:)?( )?(.+)?$").unwrap()
+});
+
+// Issue code: TDO-003
+static ISSUE_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r#"^#\s*[A-Z]{1,6}\-?\d+$"#).unwrap());
+// Issue code: 003
+static TICKET_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r#"^#\s*\d+$"#).unwrap());
+// Link to issue
+static LINK_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r#"^#\s*(http|https)://.*"#).unwrap());
+static NUM_CAPTURE_GROUPS: usize = 5usize;
+
+pub fn check_rules(tokens: &[LexResult]) -> Vec<Diagnostic> {
+    let mut diagnostics: Vec<Diagnostic> = vec![];
+    let mut prev_token_is_todo = false;
+
+    for (start, token, end) in tokens.iter().flatten() {
+        let Tok::Comment(comment) = token else {
+            continue;
+        };
+
+        let diagnostics_ref = &mut diagnostics;
+        let range = Range::new(*start, *end);
+
+        // The previous token is a TODO, so let's check if this token is a link
+        if prev_token_is_todo {
+            if ISSUE_REGEX.is_match(comment)
+                || TICKET_REGEX.is_match(comment)
+                || LINK_REGEX.is_match(comment)
+            {
+                prev_token_is_todo = false;
+                // continuing here avoids an expensive call to captures_iter() - we know that this
+                // line, if it's a link, can't be another TODO
+                continue;
+            }
+
+            diagnostics_ref.push(Diagnostic::new(MissingLinkInTodo, range));
+        }
+
+        if let Some(captures_ref) = get_captured_matches(comment).peek() {
+            let captures = captures_ref.to_owned();
+
+            // captures.get(1) is the tag, which is required for the regex to match. The unwrap()
+            // call is therefore safe
+            let tag = captures.name("tag").unwrap().as_str();
+            if tag != "TODO" {
+                diagnostics_ref.push(Diagnostic::new(
+                    InvalidTodoTag {
+                        tag: String::from(tag),
+                    },
+                    range,
+                ));
+
+                if tag.to_uppercase() == "TODO" {
+                    diagnostics_ref.push(Diagnostic::new(
+                        InvalidCapitalizationInTodo {
+                            tag: String::from(tag),
+                        },
+                        range,
+                    ));
+                }
```

Comment:
Should we use an `if-else` here to avoid two diagnostics at the same location?

```suggestion
			if tag.to_uppercase() == "TODO" {
                    diagnostics_ref.push(Diagnostic::new(
                        InvalidCapitalizationInTodo {
                            tag: String::from(tag),
                        },
                        range,
                    ));
                } else {
	                diagnostics_ref.push(Diagnostic::new(
	                    InvalidTodoTag {
	                        tag: String::from(tag),
	                    },
	                    range,
	                ));
			}   
```

Reply:
I'm not sure how heavily we frown upon multiple diagnostics in one spot - I thought it'd be fine in this instance since a "toDo", for example, also isn't the same as "TODO", so it's technically both invalid and capitalized incorrectly. Definitely open to making them exclusive though!

Reply:
On that note, @MichaReiser is it worth implementing autofixes for the rules here? I think we could do it for all of them except "missing author" and "missing link".